### PR TITLE
Update input patterns and unisonConfig for Share paths.

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

* Updates Input Patterns and arg descriptions to reflect that repos are now expected to be on share rather than git
* Threads through new Codeserver types for improved type-safety
* Adds RemoteMapping `.unisonConfig` param to specify where to pull/push for a given subpath.

```
RemoteMapping {
  me.List = "chrispenner.List"
  work.project = "git(https://github.com/work/project.git).path"
}
```

## Notes:

* Still respects old `GitUrl` mappings if no `RemoteMapping` is provided.
* Still allows pushing to git urls in `RemoteMapping` using the `git(...)` syntax.
* Fixes some git errors that were being silently swallowed

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase/Editor/RemoteRepo.hs [hunk 1]
```diff
@@ -4,36 +4,51 @@
...
+-- |
+-- >>> :set -XOverloadedLists
+-- >>> import Data.Maybe (fromJust)
+-- >>> import Network.URI
+-- >>> displayShareCodeserver DefaultCodeserver "share" ["base", "List"]
+-- "share.base.List"
+-- >>> displayShareCodeserver DefaultCodeserver "share" []
+-- "share"
+-- >>> displayShareCodeserver (CustomCodeserver . fromJust $ parseURI "https://share-next.unison-lang.org/api" >>= codeserverFromURI ) "share" []
+-- "https://share-next.unison-lang.org:443/api:.share"
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Editor/RemoteRepo.hs]: There aren't currently any parsers rigged up for this, but this is how custom shares are printed. Totally open to suggestions here.

File: parser-typechecker/src/Unison/Codebase/Path.hs [hunk 2]
```diff
@@ -92,6 +94,13 @@ newtype Path = Path {toSeq :: Seq NameSegment}
   deriving stock (Eq, Ord)
   deriving newtype (Semigroup, Monoid)
 
+-- | Meant for use mostly in doc-tests where it's
+-- sometimes convenient to specify paths as lists.
+instance GHC.IsList Path where
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Path.hs:99]: If anyone is overtly opposed to this let me know, but it definitely helps to be able to build paths tersely in doc-tests. Alternatively I'd be fine with an IsString instance,

File: unison-cli/src/Unison/Share/Types.hs [hunk 3]
```diff
@@ -1,75 +0,0 @@
-{-# LANGUAGE DeriveAnyClass #-}
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Share/Types.hs:1]: This module moved to parser-typechecker

File: unison-cli/src/Unison/Codebase/Editor/HandleInput/AuthLogin.hs [hunk 4]
```diff
@@ -1,36 +1,14 @@
...
 import Unison.Auth.OAuth (authenticateCodeserver)
 import Unison.Codebase.Editor.HandleInput.LoopState
 import Unison.Codebase.Editor.Output (Output (CredentialFailureMsg, Success))
-import Unison.Prelude
 import Unison.Share.Types
 import qualified UnliftIO
-import UnliftIO.Environment (lookupEnv)
 
--- | This is the URI where the share API is based.
-defaultShareURI :: CodeserverURI
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput/AuthLogin.hs:15]: This moved to `Unison.Share.Codeserver` since it didn't really belong here anyways.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 5]
```diff
@@ -2223,20 +2229,33 @@ resolveConfiguredUrl ::
 resolveConfiguredUrl pushPull destPath' = ExceptT do
   currentPath' <- use LoopState.currentPath
   let destPath = Path.resolve currentPath' destPath'
-  let configKey = gitUrlKey destPath
-  (eval . ConfigLookup) configKey >>= \case
-    Just url ->
-      case P.parse UriParser.writeRemotePath (Text.unpack configKey) url of
+  let remoteMappingConfigKey = remoteMappingKey destPath
+  (eval . ConfigLookup) remoteMappingConfigKey >>= \case
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:2233]: First check if the user has a `RemoteMapping`, if not, fall back on the legacy GitUrl mapping and use the deprecated parser.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 6]
```diff
@@ -1409,11 +1411,11 @@ loop = do
               case filtered of
                 [(Referent.Ref ref, ty)]
                   | Typechecker.isSubtype ty mainType ->
-                      eval (MakeStandalone ppe ref output) >>= \case
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1412]: Sorry, I'll try upgrading my ormolu again soon 🙃

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 7]
```diff
@@ -1823,14 +1825,16 @@ doPushRemoteBranch pushFlavor localPath0 syncMode = do
       let opts = PushGitBranchOpts {setRoot = True, syncMode}
       runExceptT (syncGitRemoteBranch repo opts withRemoteRoot) >>= \case
         Left gitErr -> respond (Output.GitError gitErr)
-        Right _branch -> respond Success
+        Right (Left errOutput) -> respond errOutput
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1828]: These errors were being sneakily ignored; nested either's kinda suck tbh, but they're different types here so we can't join them.

File: parser-typechecker/src/Unison/Codebase/Editor/RemoteRepo.hs (hunk 1 above)
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Editor/RemoteRepo.hs]: Where do they get printed? (I will keep reading.)
I'm thinking the Share URIs should be analogous to the Git URIs, e.g. 
`share(https://share-next.unison-lang.org/api)`
`share(https://share-next.unison-lang.org/api).share`
`share(https://share-next.unison-lang.org/api)#asfgisthash123.share` (<-- eventually, maybe)
no `:` before the path.

File: parser-typechecker/src/Unison/Codebase/Path.hs (hunk 2 above)
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Path.hs:99]: EIther/or is fine, I agree that we should limit `-XOverloadedLists` to doc-tests for example

File: parser-typechecker/src/Unison/Share/Types.hs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+codeserverIdFromCodeserverURI =
+  codeserverIdFromURIAuth . codeserverAuthority
+
+-- | Builds a servant-compatible BaseUrl for a given CodeserverURI.
+codeserverBaseURL :: CodeserverURI -> Servant.BaseUrl
+codeserverBaseURL (CodeserverURI {..}) =
+  let scheme = case codeserverScheme of
+        Https -> Servant.Https
+        Http -> Servant.Http
+      host = codeserverUserInfo <> codeserverRegName <> ":" <> show codeserverPort
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Share/Types.hs]: Does the `codeserverUserInfo` go here (into the `BaseUrl`), or does it go in into a `BasicAuth`?

File: parser-typechecker/src/Unison/Share/Types.hs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  deriving (Eq, Ord, Show)
+
+-- | This type is expanded out into all of its fields because we require certain pieces
+-- which are optional in a URI, and also to make it more typesafe to eventually convert into a
+-- BaseURL for servant clients.
+data CodeserverURI = CodeserverURI
+  { codeserverScheme :: Scheme,
+    codeserverUserInfo :: String,
+    codeserverRegName :: String,
+    codeserverPort :: Int,
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Share/Types.hs]: My gut feeling was that we should (option a) make `codeserverPort` optional, and derive it from `codeserverScheme` if it's nothing, for nicer renders; 

alternatively (option b) we could suppress it during certain renderings if it was the default; 

or (option c) just use whatever the user supplied — e.g. if they explicitly said `http://foo:80/api` then we render `http://foo:80/api`, and if they said `http://foo/api` then we render `http://foo/api`

Mitchell says he likes option b the best; I like option c the best, but haven't found a strong reason to care much between the three.

File: parser-typechecker/src/Unison/Share/Types.hs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+-- >>> import Network.URI (parseURI)
+-- >>> codeserverIdFromURI (CodeserverURI . fromJust $ parseURI "http://localhost:5424/api")
+-- >>> codeserverIdFromURI (CodeserverURI . fromJust $ parseURI "https://share.unison-lang.org/api")
+-- Right "localhost"
+-- Right "share.unison-lang.org"
+codeserverIdFromURI :: URI -> Either Text CodeserverId
+codeserverIdFromURI uri =
+  case uriAuthority uri of
+    Nothing -> Left $ "No URI Authority for URI " <> tShow uri
+    Just ua -> pure $ codeserverIdFromURIAuth ua
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Share/Types.hs:127]: I know we'd talked about it briefly, but shouldn't we be using the whole `CodeserverURI` in the `CodeserverId`?

Edit: Ok, right, it was because it was going to be O(path segments) to find the application base path to look up the token key. 

But in that case, after a non-authenticated request, do we have the same problem coming up with a path to start the authentication process with?

e.g. `https://share.unison-lang.org/api/sync/path/get`
becomes `https://share.unison-lang.org/api`
becomes `https://share.unison-lang.org/api/.openid/well-known-configuration`

File: parser-typechecker/src/Unison/Share/Types.hs [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+-- | Gets the part of the CodeserverURI that we use for identifying that codeserver in
+-- credentials files.
+--
+-- >>> import Data.Maybe (fromJust)
+-- >>> import Network.URI (parseURI)
+-- >>> codeserverIdFromURI (CodeserverURI . fromJust $ parseURI "http://localhost:5424/api")
+-- >>> codeserverIdFromURI (CodeserverURI . fromJust $ parseURI "https://share.unison-lang.org/api")
+-- Right "localhost"
+-- Right "share.unison-lang.org"
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Share/Types.hs:122]: It looks like these need to be re-run, because it looks like the function accepted a different type at the time, compared to now.

File: parser-typechecker/src/Unison/Share/Types.hs (hunk 10 above)
Comment (aryairani) [reply parser-typechecker/src/Unison/Share/Types.hs:127]: Is this still used, or should we just be using `codeserverIdFromCodeserverURI`?

File: parser-typechecker/src/Unison/Share/Types.hs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+codeserverIdFromURI :: URI -> Either Text CodeserverId
+codeserverIdFromURI uri =
+  case uriAuthority uri of
+    Nothing -> Left $ "No URI Authority for URI " <> tShow uri
+    Just ua -> pure $ codeserverIdFromURIAuth ua
+
+-- | Builds a CodeserverId from a URIAuth
+codeserverIdFromURIAuth :: URIAuth -> CodeserverId
+codeserverIdFromURIAuth ua =
+  (CodeserverId (Text.pack $ uriUserInfo ua <> uriRegName ua <> uriPort ua))
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Share/Types.hs:132]:
```suggestion
  CodeserverId (Text.pack $ uriUserInfo ua <> uriRegName ua <> uriPort ua)
```

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs (hunk 7 above)
Comment (aryairani) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1828]: Good catch.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 13]
```diff
@@ -2223,20 +2229,33 @@ resolveConfiguredUrl ::
...
+            Right ns ->
+              pure . Right $ ns
+        Nothing ->
+          pure . Left $ NoConfiguredRemoteMapping pushPull destPath
+    Just url -> do
+      case P.parse UriParser.writeRemotePath (Text.unpack remoteMappingConfigKey) url of
         Left e ->
           pure . Left $
-            ConfiguredGitUrlParseError pushPull destPath' url (show e)
+            ConfiguredRemoteMappingParseError pushPull destPath url (show e)
```
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:2251]: I'll look for what this output looks like; I'm wondering if it will say which config section the error is in?

File: unison-cli/src/Unison/Codebase/Editor/UriParser.hs [hunk 14]
```diff
@@ -87,26 +88,28 @@ readShareRemoteNamespace :: P ReadShareRemoteNamespace
...
 readGitRemoteNamespace :: P ReadGitRemoteNamespace
 readGitRemoteNamespace = P.label "generic git repo" $ do
   P.string "git("
+  P.space
   protocol <- parseGitProtocol
   treeish <- P.optional gitTreeishSuffix
+  P.space
   let repo = ReadGitRepo {url = printProtocol protocol, ref = treeish}
   P.string ")"
   nshashPath <- P.optional namespaceHashPath
```
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/UriParser.hs]: Why this change?

File: unison-cli/src/Unison/Codebase/Editor/UriParser.hs [hunk 15]
```diff
@@ -297,7 +350,7 @@ nameSegment =
 gitTreeishSuffix :: P Text
 gitTreeishSuffix = P.label "git treeish" . P.try $ do
   void $ C.char ':'
-  P.takeWhile1P (Just "not close paren") (/= ')')
+  P.takeWhile1P (Just "not close paren") (\c -> c /= ')' && not (isSpace c))
```
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/UriParser.hs]: Was this to support whitespaces inside the parens in the config file, or for something else?

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs (hunk 13 above)
Comment (aryairani) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:2251]: It looks like you could have an error in your `GitUrl` section and UCM would tell you you have an error in your `RemoteMapping` section; I think we should have separate errors for each, even though it's more typing. (Or it could be parameterized.)

File: unison-cli/src/Unison/CommandLine/InputPatterns.hs [hunk 16]
```diff
@@ -1086,9 +1086,9 @@ push =
                 "publishes the current namespace into the remote namespace `remote`"
               ),
               ( "`push`",
-                "publishes the current namespace"
-                  <> "into the remote namespace configured in `.unisonConfig`"
-                  <> "with the key `GitUrl.ns` where `ns` is the current namespace"
+                "looks up the remote namespace configured in `.unisonConfig`"
+                  <> "at the key `RemoteMappings.<namespace>` where `<namespace>` is the current namespace,"
+                  <> "then publishes the current namespace to that location."
```
Comment (aryairani) [inline unison-cli/src/Unison/CommandLine/InputPatterns.hs]: I think we want to keep the help text for `push` to have its emphasis on "pushing"/"publishing" (like before) vs emphasizing the remote namespace lookup. — Same for `pull`, above.

Comment (aryairani) [reply unison-cli/src/Unison/CommandLine/InputPatterns.hs]: "publishes the current namespace into its associated remote namespace in .unisonConfig"

or
"publishes the current namespace into the remote namespace configured in `.unisonConfig` at the key `RemoteMappings.<namespace>` where `<namespace>` is the current namespace.

We should probably add a separate help page and/or help topic for `remote-mappings`, but that can be a separate ticket.

File: unison-cli/src/Unison/CommandLine/OutputMessages.hs [hunk 17]
```diff
@@ -1657,7 +1658,7 @@ notifyUser dir o = case o of
     _nameChange _cmd _pastTenseCmd _oldName _newName _r = error "todo"
     expectedNonEmptyPushDest writeRemotePath =
       P.lines
-        [ P.wrap ("The remote namespace" <> prettyWriteRemotePath writeRemotePath <> "is empty."),
+        [ P.wrap ("The remote namespace " <> prettyWriteRemotePath writeRemotePath <> " is empty."),
```
Comment (aryairani) [inline unison-cli/src/Unison/CommandLine/OutputMessages.hs:1661]: I think that due to the way `P.wrap` works that this change doesn't actually affect the output.

File: parser-typechecker/src/Unison/Codebase/Editor/RemoteRepo.hs (hunk 1 above)
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Editor/RemoteRepo.hs]: I kept reading but I forgot to notice where they get printed. Though since I don't think we have parsers for these yet (before this PR or in it), I guess it doesn't matter yet?

Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Codebase/Editor/RemoteRepo.hs]: They're used in error messages, but no we don't have parsers for these as arguments yet, so they won't come up for users unless they pass the `UNISON_SHARE_HOST` env var

File: parser-typechecker/src/Unison/Share/Types.hs (hunk 8 above)
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Share/Types.hs]: This is how it would appear when provided, but that's a good a point.

Moving it to BasicAuth would require re-writing pieces of the Authenticated Client middleware; so I vote we leave it like this until anyone actually needs it, e.g. for a private unison-cloud or something, but I think we'd prefer they didn't use BasicAuth for that 🙃

File: parser-typechecker/src/Unison/Share/Types.hs (hunk 9 above)
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Share/Types.hs]: I'll tie-break and vote `c` as well 👍🏼

File: unison-cli/src/Unison/Codebase/Editor/UriParser.hs (hunk 15 above)
Comment (ChrisPenner) [reply unison-cli/src/Unison/Codebase/Editor/UriParser.hs]: Yes, I was under the impression it would be more user-friendly, but seeing as how these will typically be provided as ucm cli args I suppose it's not possible to have spaces in a single arg anyways; I'll roll it back.

File: unison-cli/src/Unison/CommandLine/OutputMessages.hs (hunk 17 above)
Comment (ChrisPenner) [reply unison-cli/src/Unison/CommandLine/OutputMessages.hs:1661]: I'd added the spaces before a merge which added the wrap 🙃

File: parser-typechecker/src/Unison/Share/Types.hs (hunk 10 above)
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Share/Types.hs:127]: I'll address this in an upcoming PR which handles refresh tokens 👍🏼
